services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -301,6 +301,54 @@ data class InvoicingPostalAddressRequest(
     val city: String?,
 )
 
+@JvmName("hasChangesInKaivuilmoitusAlueList")
+fun List<KaivuilmoitusAlue>?.hasChanges(areas: List<KaivuilmoitusAlue>?): Boolean {
+    if (this == null) {
+        return areas != null
+    }
+    if (areas == null) {
+        return true
+    }
+    if (this.size != areas.size) {
+        return true
+    }
+    return this.any { it.hasChanges(areas.find { area -> area.hankealueId == it.hankealueId }) }
+}
+
+fun KaivuilmoitusAlue.hasChanges(area: KaivuilmoitusAlue?): Boolean {
+    if (area == null) {
+        return true
+    }
+    return this.hankealueId != area.hankealueId ||
+        this.tyoalueet.hasChanges(area.tyoalueet) ||
+        this.katuosoite != area.katuosoite ||
+        this.tyonTarkoitukset != area.tyonTarkoitukset ||
+        this.meluhaitta != area.meluhaitta ||
+        this.polyhaitta != area.polyhaitta ||
+        this.tarinahaitta != area.tarinahaitta ||
+        this.kaistahaitta != area.kaistahaitta ||
+        this.kaistahaittojenPituus != area.kaistahaittojenPituus ||
+        this.lisatiedot != area.lisatiedot ||
+        this.name != area.name
+}
+
+@JvmName("hasChangesInTyoalueList")
+fun List<Tyoalue>.hasChanges(workAreas: List<Tyoalue>): Boolean {
+    if (this.size != workAreas.size) {
+        return true
+    }
+    for (i in this.indices) {
```

Comment:
Here we assume that if the order of tyoalue differes, then it can be marked as a change.

Reply:
If the list ordering matters, it should matter everywhere.

Reply:
I got rid off these by overloading equals and hashCode in Tyoalue.